### PR TITLE
Incremental PCA

### PR DESCRIPTION
**Pull request by ordabayevy**

### CODE REVIEW

**Comment by ordabayevy** [comment]
Blocked by #60

File: examples/incremental_pca.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+       (https://www.cs.toronto.edu/~dross/ivt/RossLimLinYang_ijcv.pdf)
+"""
+
+from lightning.pytorch.cli import LightningCLI
+
+from scvid.data import DistributedAnnDataCollectionDataModule
+from scvid.train.training_plan import TrainingPlan
+
+
+class IncrementalPCACLI(LightningCLI):
```
Comment (mbabadi) [inline examples/incremental_pca.py:36]: The naming clash between `IncrementalPCAFromCli` in `scvid/module` and `IncrementalPCACLI` here makes the code difficult to read. How about renaming `IncrementalPCACLI` in here (`examples/incremental_pca.py`) to `_LightningCLIWithLinks`?

File: scvid/module/from_cli.py [hunk 2]
```diff
@@ -94,3 +95,37 @@ def __init__(
             seed=seed,
             transform=transform,
         )
+
+
+class IncrementalPCAFromCli(IncrementalPCA):
```
Comment (mbabadi) [inline scvid/module/from_cli.py:100]: `IncrementalPCAFromCli` -> `IncrementalPCAFromCLI` (here and elsewhere)

File: scvid/module/incremental_pca.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            If ``False`` then the data is assumed to be centered and the mean correction
+            is not applied to the update step.
+    """
+
+    def __init__(
+        self,
+        g_genes: int,
+        k_components: int,
+        p_oversamples: int = 10,
+        mean_correct: bool = False,
```
Comment (mbabadi) [inline scvid/module/incremental_pca.py:41]: `mean_correct` -> `perform_mean_correction`

File: scvid/module/incremental_pca.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    def forward(self, x_ng: torch.Tensor) -> None:
+        if self.transform is not None:
+            x_ng = self.transform(x_ng)
+
+        g = self.g_genes
+        k = self.k_components
+        p = self.p_oversamples
+        m = self.x_size
+        n = x_ng.size(0)
+        assert p + k <= min(n, g)
```
Comment (mbabadi) [inline scvid/module/incremental_pca.py:76]: Please write a user-readable exception message for failing asserts.

File: scvid/module/incremental_pca.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        if self.mean_correct:
+            x_mean_g = x_ng.mean(dim=0)
+            _, S_q, V_gq = torch.svd_lowrank(x_ng - x_mean_g, q=k + p)
+        else:
+            _, S_q, V_gq = torch.svd_lowrank(x_ng, q=k + p)
+
+        # if not the first batch, merge results
+        if m > 0:
+            SV_kg = torch.diag(S_q[:k]) @ V_gq.T[:k]
+            C = torch.cat([torch.diag(self.S_k) @ self.V_kg, SV_kg], dim=0)
```
Comment (mbabadi) [inline scvid/module/incremental_pca.py:88]: I believe `torch.einsum("k,kg->kg", self.S_k, self.V_kg)` is faster than `torch.diag(self.S_k) @ self.V_kg`.

File: scvid/module/incremental_pca.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            if self.mean_correct:
+                mean_correction = (
+                    math.sqrt(m * n / (m + n)) * (self.x_mean_g - x_mean_g)[None, :]
+                )
+                C = torch.cat([C, mean_correction], dim=0)
+            # perform SVD on merged results
+            _, S_q, V_gq = torch.svd_lowrank(C, q=k + p)
+
+        # update buffers
+        self.V_kg = V_gq.T[:k]
```
Comment (mbabadi) [inline scvid/module/incremental_pca.py:98]: Hmmm... shouldn't with actually be keeping `self.V_qg` and truncate at the very end? I feel like there is something wrong about eager truncation.  With eager truncation, we've effectively working in a k-dim space except for when we're doing the CVD on C, whereas we could be working in a q-dim space by keeping track of oversampled PCs all the way through (but only keeping the first k at the very end).

File: test/test_ipca.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    rng.manual_seed(1465)
+    mean_g = torch.randn((g,), generator=rng)
+    x_ng = torch.randn((n, g), generator=rng) + mean_g
+    return x_ng
+
+
+@pytest.mark.parametrize("mean_correct", [False, True])
+@pytest.mark.parametrize("batch_size", [10_000, 5000, 1000, 500, 250])
+@pytest.mark.parametrize("k", [30, 50, 80])
+def test_incremental_pca(x_ng: np.ndarray, mean_correct: bool, batch_size: int, k: int):
```
Comment (mbabadi) [inline test/test_ipca.py:29]: Don't we want to check `p_oversamples`, i.e. that the error decreases with increased `p_oversamples`?  Also, how about checking the results against exact SVD?  Neither is necessary but they'd make good tests.

**Review by mbabadi — CHANGES_REQUESTED** [review]
Nicely done! :)

File: scvid/module/incremental_pca.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        n = x_ng.size(0)
+        assert k <= min(n, g), (
+            f"Rank of svd_lowrank (k_components): {k}"
+            f" must be less than min(n_cells, g_genes): {min(n, g)}"
+        )
+
+        # compute SVD of new data
+        if self.perform_mean_correction:
+            x_mean_g = x_ng.mean(dim=0)
+            _, S_k, V_gk = torch.svd_lowrank(x_ng - x_mean_g, q=k)
```
Comment (mbabadi) [inline scvid/module/incremental_pca.py:79]: Please expose `niter` and make it kwarg of `IncrementalPCA` and call it `svd_lowrank_niter`.  Also, expose it to CLI.  No need to test it.

**Review by mbabadi — CHANGES_REQUESTED** [review]

File: scvid/module/incremental_pca.py (hunk 8 above)
Comment (ordabayevy) [reply scvid/module/incremental_pca.py]: Done!

**Review by mbabadi — APPROVED** [review]
